File: app/src/main/java/com/narcyber/mvpbasics/presenter/ParentPresenter.java
package com.narcyber.mvpbasics.presenter;

import com.narcyber.mvpbasics.model.User;
import com.narcyber.mvpbasics.repository.UserRepository;

import java.util.List;

import io.reactivex.rxjava3.disposables.CompositeDisposable;

public abstract class ParentPresenter implements UserRepository.PresenterContract {

    private final CompositeDisposable compositeDisposable;
    private final UserRepository userRepository;

    public ParentPresenter() {
        this.compositeDisposable = new CompositeDisposable();
        this.userRepository = new UserRepository(this);
    }

    public UserRepository getUserRepository() {
        return this.userRepository;
    }

    @Override
    public void respondSuccessPushing() {
    }

    @Override
    public void respondPushingFailed(String message) {
    }

    @Override
    public void respondUserByEmail(User user) {
    }

    @Override
    public void respondUserByUserName(User user) {
    }

    @Override
    public void respondUserByEmailAndPassword(User user) {
    }

    @Override
    public void respondDeletedUserCount(int count) {
    }

    @Override
    public void respondFailedDeleteUser(String message) {
    }

    @Override
    public void respondForAllUsers(List<User> users) {
    }


    @Override
    public CompositeDisposable getDisposables() {
        return this.compositeDisposable;
    }

    public void clearDisposables() {
        this.compositeDisposable.clear();
    }
}

File: app/src/main/java/com/narcyber/mvpbasics/presenter/ParentPresenter.kt
package com.narcyber.mvpbasics.presenter

import com.narcyber.mvpbasics.model.User
import com.narcyber.mvpbasics.repository.UserRepository
import io.reactivex.rxjava3.disposables.CompositeDisposable


abstract class ParentPresenter : UserRepository.PresenterContract {
    private val compositeDisposable: CompositeDisposable by lazy {
        CompositeDisposable()
    }
    val userRepository: UserRepository by lazy {
        UserRepository(this)
    }
    override val disposables: CompositeDisposable?
        get() = compositeDisposable

    override fun respondSuccessPushing() {
    }

    override fun respondPushingFailed(message: String?) {
    }

    override fun respondUserByEmail(user: User?) {
    }

    override fun respondUserByUserName(user: User?) {
    }

    override fun respondUserByEmailAndPassword(user: User?) {
    }

    override fun respondDeletedUserCount(count: Int) {
    }

    override fun respondFailedDeleteUser(message: String?) {
    }

    override fun respondForAllUsers(users: List<User?>?) {
    }

    open fun clearDisposables() {
        compositeDisposable.clear()
    }

}